### PR TITLE
[bug-fix] Fixing target and prediction mapping for eval_fn

### PR DESCRIPTION
**Pull request by sunishsheth2009**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/sunishsheth2009/mlflow/pull/10002?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/10002/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 10002
```

</p>
</details>

### What changes are proposed in this pull request?
[bug-fix] Fixing target and prediction mapping for eval_fn

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/evaluate/test_default_evaluator.py [hunk 1]
```diff
@@ -2927,3 +2928,91 @@ def test_evaluate_with_latency():
...
+        "input_tokens": 569,
+        "output_tokens": 93,
+        "total_tokens": 662,
+        "model": "gpt-3.5-turbo-0613",
+        "route_type": "llm/v1/completions",
+    },
+}
+
+
+def test_evaluate_with_correctness():
```
Comment (dbczumar) [inline tests/evaluate/test_default_evaluator.py:2950]: I tried this code on `master` just to confirm that it reproduces the issue, and it does. While repro-ing the issue, I saw this message: 

```
Below are the existing column names for the input/output data:
Input Columns: ['inputs']
Output Columns: []
Note that this does not include the output column: 'output'
```

The part about
```
Note that this does not include the output column: 'output'
```

is kind of confusing. Why not include `output` in the `Output columns` in the error message? It is an output column :)

cc @annzhang-db @sunishsheth2009

Comment (dbczumar) [reply tests/evaluate/test_default_evaluator.py:2950]: @annzhang-db @sunishsheth2009 it looks like the name of the output column in the error message is incorrect too. For example, I ran the following code with `ground_truthz` (intentional typo) as the grading context column. The error message mentioned the output column `output`, but the output column is actually called `predictions`.

```
import mlflow.metrics
import mlflow
import pandas as pd

import os
os.environ["OPENAI_API_KEY"] = "<REDACTED>"

metric = mlflow.metrics.make_genai_metric(
      name="correctness",
      definition=(
          "Correctness refers to how well the generated output matches "
          "or aligns with the reference or ground truth text that is considered "
          "accurate and appropriate for the given input. The ground truth serves as "
          "a benchmark against which the provided output is compared to determine the "
          "level of accuracy and fidelity."
      ),
      grading_prompt=(
          "Correctness: If the answer correctly answer the question, below "
          "are the details for different scores: "
          "- Score 0: the answer is completely incorrect, doesn’t mention anything about "
          "the question or is completely contrary to the correct answer. "
          "- Score 1: the answer provides some relevance to the question and answer "
          "one aspect of the question correctly. "
          "- Score 2: the answer mostly answer the question but is missing or hallucinating "
          "on one critical aspect. "
          "- Score 4: the answer correctly answer the question and not missing any "
          "major aspect"
      ),
      examples=[],
      version="v1",
      model="openai:/gpt-4",
      grading_context_columns=["ground_truthz"],
      parameters={"temperature": 0.0},
      aggregations=["mean", "variance", "p90"],
      greater_is_better=True,
  )

with mlflow.start_run():
    eval_df = pd.DataFrame(
        {
            "inputs": [
                "What is MLflow?",
                "What is Spark?",
                "What is Python?",
            ],
            "ground_truth": [
                "MLflow is an open-source platform",
                "Apache Spark is an open-source, distributed computing system",
                "Python is a high-level programming language",
            ],
            "prediction": [
                "MLflow is an open-source platform",
                "Apache Spark is an open-source, distributed computing system",
                "Python is a high-level programming language",
            ],
        }
    )
    results = mlflow.evaluate(
        data=eval_df,
        evaluators="default",
        targets="ground_truth",
        predictions="prediction",
        extra_metrics=[metric],
    )

    print(results.metrics)
    print(results.tables["eval_results_table"].to_dict().keys())
```

Comment (dbczumar) [reply tests/evaluate/test_default_evaluator.py:2950]: @annzhang-db @sunishsheth2009 I also found that `col_mapping` is broken with `targets`. The following should work, but it doesn't:

```
import mlflow.metrics
import mlflow
import pandas as pd

import os
os.environ["OPENAI_API_KEY"] = "<REDACTED>"

metric = mlflow.metrics.make_genai_metric(
      name="correctness",
      definition=(
          "Correctness refers to how well the generated output matches "
          "or aligns with the reference or ground truth text that is considered "
          "accurate and appropriate for the given input. The ground truth serves as "
          "a benchmark against which the provided output is compared to determine the "
          "level of accuracy and fidelity."
      ),
      grading_prompt=(
          "Correctness: If the answer correctly answer the question, below "
          "are the details for different scores: "
          "- Score 0: the answer is completely incorrect, doesn’t mention anything about "
          "the question or is completely contrary to the correct answer. "
          "- Score 1: the answer provides some relevance to the question and answer "
          "one aspect of the question correctly. "
          "- Score 2: the answer mostly answer the question but is missing or hallucinating "
          "on one critical aspect. "
          "- Score 4: the answer correctly answer the question and not missing any "
          "major aspect"
      ),
      examples=[],
      version="v1",
      model="openai:/gpt-4",
      grading_context_columns=["ground_truthz"],
      parameters={"temperature": 0.0},
      aggregations=["mean", "variance", "p90"],
      greater_is_better=True,
  )

with mlflow.start_run():
    eval_df = pd.DataFrame(
        {
            "inputs": [
                "What is MLflow?",
                "What is Spark?",
                "What is Python?",
            ],
            "ground_truth": [
                "MLflow is an open-source platform",
                "Apache Spark is an open-source, distributed computing system",
                "Python is a high-level programming language",
            ],
            "prediction": [
                "MLflow is an open-source platform",
                "Apache Spark is an open-source, distributed computing system",
                "Python is a high-level programming language",
            ],
        }
    )
    results = mlflow.evaluate(
        data=eval_df,
        evaluators="default",
        targets="ground_truthz",
        predictions="prediction",
        extra_metrics=[metric],
        evaluator_config={
            "col_mapping": {
                "ground_truthz": "ground_truth"
            }
        }
    )

    print(results.metrics)
    print(results.tables["eval_results_table"].to_dict().keys())
```

**Review by dbczumar — APPROVED** [review]
@sunishsheth2009 Awesome fix! LGTM once we're tracking / prioritizing resolution of some related issues discovered while testing the PR. cc @annzhang-db @prithvikannan

File: tests/evaluate/test_default_evaluator.py (hunk 1 above)
Comment (sunishsheth2009) [reply tests/evaluate/test_default_evaluator.py:2953]: Case 1 and 2 should be resolved in this PR: https://github.com/mlflow/mlflow/pull/9983#pullrequestreview-1686585155
And adding solution for case 3 in this PR #9705

**Review by prithvikannan — APPROVED** [review]
LGTM!

**Pull request merged**